Configurabe owncloud temporary directory

mmattel opened the pull request:
This is a PR for #12910

It is now possible to make the temporary directory used by owncloud configurable via config.php

The tag used is: 'tempdirectory' 

If this parameter does not exist, all stays the same as before.
If the parameter exists, the given path is used if the path is physically present and writable by owncloud.

I have tried all possibilities:
- having no tag present
- tag present but path contains special/invalid characters
- tag and path present but path is physically not present
- tag and path present, path exists physically but is not writable (wrong permissions)

I also added two debug log entries:
- log which temporary path is used by owncloud (in any case)
- log if the config.php path given is not present / not writable

If this PR gets approved and merged, I would also do a PR for the documentation.

Marry Christmas to you all

mmattel commented:
Question: I made a PR in documentation to describe the background and a solution  (https://github.com/owncloud/documentation/issues/727, pls see also my reference above) including a note in the PR that the documentation part should get merged only after this PR is merged, especially because I reference to the new parameter ``tempdirectory`` which is introduced by this PR. But the documentation got merged yesterday, maybe because my note about the dependency was overseen.
Please advice how to proceed.

Xenopathic commented:
I like the idea, but I can see certain operations resulting in a *lot* of debug messages about the temp dir being spewed to the log. Perhaps use a lazy initialisation of a variable (during which a log entry can be written), then just use that 'cached' variable for all future `get_temp_dir()` calls.

In addition, I would say this should really be part of `OC::Server`. In fact, putting it there would make the code nice and clean with regard to lazy initialization. Just register another service and away you go.

mmattel commented:
Regarding the quantity of debug messages beeing logged:

karlitschek commented this already and I changed the code that logging only takes place in following circumstances (pls see commits above):

1. if the ``config.php`` parameter ``loglevel``is set to 0 (debug)
2. if there is the parameter ``tempdirectory``present but the directory can not be accessed or there is no write access granted

In normal owncloud operations, a admin will change the loglevel parameter to eg warning to not pollute the log anyway - first problem solved.
And when you set the tempdirectory parameter and something goes wrong, I must know this to get warned - second problem solved.

Regarding all the other things commented, I have to admint that my PHP knowledge is too low to handle that...

MorrisJobke commented:
@mmattel 0 changes? I guess there is a problem with the back and forth. Can you have a look?

mmattel commented:
Fixed, pls take a look now.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/7c06dd7c-3fdf-43fd-af58-8fa393f9bd63) was created.

mmattel commented:
7b55c32 is a squash of all 5 commits, for better read/usability

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/7c06dd7c-3fdf-43fd-af58-8fa393f9bd63) completed: **4 new issues**

Xenopathic commented on lib/base.php:
```diff
@@ -978,10 +978,31 @@ protected static function tryFormLogin() {
+				$text = 'NULL';
+			}
+			\OCP\Util::writeLog('Core', 'Temporary directory set to: ' . $text, \OCP\Util::DEBUG);
+		}
```
This whole block can be squashed down to a single line, removing the `if` statement:

    \OCP\Util::writeLog('Core', 'Temporary directory set to: ' . ($text ? $text : 'NULL'), \OCP\Util::DEBUG);

Xenopathic commented on lib/base.php:
```diff
@@ -992,9 +1013,29 @@ function get_temp_dir() {
+			// necessary catch if try fails
+		}	catch (Exception $e) {
+				return false;
+			}
```
Looks like the formatting around the `catch` is a little wacky :laughing:.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/b22bf78e-da5c-4d89-a569-0af112614c80) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/59551995-b9bc-4cbc-9bd1-c5651ea4becc) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/fb1814be-48e9-4a9a-9936-e60052797165) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/b22bf78e-da5c-4d89-a569-0af112614c80) completed: **4 new issues**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/59551995-b9bc-4cbc-9bd1-c5651ea4becc) completed: **4 new issues**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/fb1814be-48e9-4a9a-9936-e60052797165) completed: **4 new issues**

Xenopathic commented:
The logging code won't write a log entry if `loglevel` is higher than the log level passed during the call to write a log entry, so the `if` guard around that log write is unnecessary. I also noticed that the formatting around the `catch` is still wacky? The `catch` should be on the same line as the closing brace of `try`, and the indentation inside the `catch` should only be one level deeper, not two as it is at the moment.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/d8b0545a-930a-4a09-aa17-d46d9e9cdd91) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/d8b0545a-930a-4a09-aa17-d46d9e9cdd91) completed: **6 new issues**

MorrisJobke commented:
@mmattel we are now in feature freeze and this has to wait for ownCloud 8.1 as it is a feature

Xenopathic commented:
When merges to master are accepted again, this gets a :+1: from me

nickvergessen commented on lib/base.php:
```diff
@@ -992,9 +1006,28 @@ function get_temp_dir() {
+	 * @return bool|string Path to the temporary directory or false
+	 */
+	function get_config_temp_dir() {
+		$temp = \OC::$server->getConfig()->getSystemValue('tempdirectory', false);
```
I'd return here, when `!$temp`
This helps ignoring empty strings aswell as prevents your code below from testing `is_writeable(false)`, etc.

mmattel replied:
But ``!$temp`` does not check if the directory has proper permisions to write to.
is_writable checks if the directory is valid, accessible AND writable - what is definitely needed for temp space.
If we would return as suggested, nothing would get logged and nobody would see the problem.
I tried this code with many thinkable directories. Valid one, once containing bad characters, empty, no write access and well working once. The important thing is, that in case of problems, these get logged and the code continues to run - mission critical.
But please correct me and do a suggestion for a better approach if I have overseen something.

nickvergessen replied:
I meant something like:
```php
if (!$temp) {
	return false;
}
```

nickvergessen commented on lib/base.php:
```diff
@@ -978,10 +978,24 @@ protected static function tryFormLogin() {
+	 * If a temporary directory is set in config.php, use this one
+	 * @return null|string Path to the temporary directory or null
+	 */
+	function gather_temp_dir() {
```
Should these functions be in their own `if (!function_exists('***')) {`?

mmattel replied:
``get_temp_dir`` is the entry point for all the other functions and queried via ``!function_exists`` once properly, why should we query this on each function again?

LukasReschke commented:
@owncloud-bot Retest this please.

LukasReschke commented:
@Xenopathic https://github.com/owncloud/core/pull/13031#issuecomment-68702353? ;-)

Xenopathic commented:
:+1: stands

MorrisJobke commented:
@nickvergessen @LukasReschke What is the state of this? ready to merge?

DeepDiver1975 commented:
fixes #14886

DeepDiver1975 commented:
@mmattel please squash your commits - THX

mmattel committed:
configurabe owncloud temporary directory

added loglevel dependency for logging

added the needed description in config.sample.php

changed log type to warn when path causes problems

changes as requested, some small improvements

adoptions/changes as requested

optical stuff

remove if statement and cleared formatting at catch

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/09b9d724-c924-4ee2-8cb5-758237050ab6) was created.

mmattel commented:
squashed

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/09b9d724-c924-4ee2-8cb5-758237050ab6) completed: **3 new issues**

Xenopathic commented:
Test failure due to known bug with timezone: #15094 - all other tests pass.

For the record, my :+1: still holds. cc @DeepDiver1975 @PVince81 @LukasReschke nice easy review?

nickvergessen commented:
@owncloud-bot retest this please

MorrisJobke commented:
@DeepDiver1975 @icewind1991 Is this a valid usage?

icewind1991 commented on lib/base.php:
```diff
@@ -978,10 +978,24 @@ protected static function tryFormLogin() {
 
 if (!function_exists('get_temp_dir')) {
```
Take this opportunity to move this to a proper place?

icewind1991 replied:
(and not have them as non-class functions)

mmattel replied:
This code line was also present in the original code, which is currently line 1022. I took it and did not changed it.

icewind1991 commented on lib/base.php:
```diff
@@ -978,10 +978,24 @@ protected static function tryFormLogin() {
+		// Get the temporary directory and log the path if loglevel is set to debug
+		// Info: based on the temp dir, further directories may be created unique to the instance
+		$temp = gather_temp_dir();
+		\OCP\Util::writeLog('Core', 'Temporary directory set to: ' . ($temp ? $temp : 'NULL'), \OCP\Util::DEBUG);
```
I would prefer not to spam the log with debug messages for each request

mmattel replied:
As from my understanding, debug messages are only written when the debug flag is set, so this does not pollute the standard operating instance log and debugging mode is non standard operating.

If you compare other code elements, especially from gallery ect, you have tons of debug messages written compared to this PR. tempdirectory does not get that often accessed compared with eg gallery.

Please keep in mind, that having NO logging means no hint where temp data will be written to what is somehow counteracting when you do debugging.

But agree, there should be in ganeral a better approach for managing the tempdirectory. This because each request for the directory location is passing this code. This should be happening only once at the beginning, but this is then a different story.

If you agree, I propose to merge this PR to have the functionality added and then do a design Q&A how to improve this internally.

mmattel commented:
@icewind1991 
could I request for a thumb up to get that merged?

mmattel commented:
Any news on that? Would be happy if we can get that merged.

icewind1991 commented:
Can you move the code to https://github.com/owncloud/core/blob/master/lib/private/tempmanager.php

Xenopathic commented:
@icewind1991 Hmm, that'd certainly make more sense. But that would probably mean delaying this to 8.2...

PVince81 commented:
Note that we are past feature freeze so only bugfixes are allowed in. 8.2 would probably fit better then, especially that it touches some of the core parts.

MorrisJobke commented:
@mmattel As we are in feature freeze this sadly has to wait for 8.2. Sorry for that :(

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/d3bab3bd-d34d-4da0-a282-6c134a3a52fd) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/859031ca-f9aa-4c1c-9448-f709fdb16aef) was created.

mmattel committed:
moved code to /lib/private/tempmanager.php

fix for unit test

some fixes

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/d0c5c924-7491-4b46-87f0-b7db3b03b984) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/3b26700d-54ab-4f85-9280-84f055f079bd) was created.

mmattel commented:
Closing due to rebase problems, will reopen in an new PR
